Modules/arraymodule.c

```diff
@@ -2849,17 +3251,21 @@ array_new(PyTypeObject *type, PyObject *args, PyObject *kwds)
                     }
 
                     arrayobject *self = (arrayobject *)a;
-                    // self->ob_item may be NULL but it is safe.
-                    PyMem_Free(self->ob_item);
-                    self->ob_item = (char *)ustr;
+                    self->data = arraydata_set_items(self->data,
+                        (char *)ustr, n, sizeof(Py_UCS4), false);
+                    PyMem_Free(ustr);
+                    if (self->data == NULL) {
+                        Py_DECREF(a);
+                        PyErr_NoMemory();
+                        return NULL;
+                    }
                     Py_SET_SIZE(self, n);
-                    self->allocated = n;
                 }
             }
             else if (initial != NULL && array_Check(initial, state) && len > 0) {
                 arrayobject *self = (arrayobject *)a;
                 arrayobject *other = (arrayobject *)initial;
-                memcpy(self->ob_item, other->ob_item, len * other->ob_descr->itemsize);
+                memcpy(array_items_ptr(self), other->data->items, len * other->ob_descr->itemsize);
```

Comment:
Most of the uses of `array_items_ptr()` don't seem necessary. 
`self->data->items` can't be NULL here. This can be:

```
memcpy(self->data->items, other->data->items, len * other->ob_descr->itemsize);
```

Reply:
`array_items_ptr()` seems useful if you need to handle the case where `self->data` is NULL. But if `self->data` isn't NULL, then it seems clearer to write `self->data->items` to me.

In this specific example, the `array_items_ptr(self)` vs. `other->data->items` stuck out. Neither can be a NULL pointer (or we have a bug).

Reply:
Can be NULL pointers to memcpy if copy size is 0, probably not officially though, but original array module code counted on this behavior as well.

Reply:
`array_items_ptr()` exists because of:
```
return PyBytes_FromStringAndSize(array_items_ptr(self),
                            Py_SIZE(self) * self->ob_descr->itemsize);
```
Which treats NULL as a special value and so couldn't be `array->data->items` if `data == NULL` because that would give `(void *)8` which is not NULL.

Reply:
In any case, I know where it is needed and not and can remove where not or leave harmlessly consistent, your call.